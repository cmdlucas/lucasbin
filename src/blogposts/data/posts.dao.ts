import { readPostsContent } from "./posts.reader";
import { getMarkdownData, Markdown } from "../../shared/tool/markdown";
import path from "path";
import imageToBase64 from "base64-img";
import slugify from "slugify";

export interface Post extends Markdown {
  summary: string;
  header_image: string;
  pid: string;
}

export interface PidToPostMap {
  [x: string]: number;
}

const seoUrl = (title: string) => {
  const strippedTitle = title
    .toLowerCase()
    .replace(/and/g, "")
    .replace(/ a /g, " ")
    .replace(/the/g, "")
    .trim();
  return slugify(strippedTitle, {
    lower: true,
    strict: true,
    replacement: "-",
  });
};

export const getMarkdown = (): Markdown[] => {
  return getMarkdownData(readPostsContent());
};

export const getPostSummary = (markdown: Markdown): string => {
  return `${markdown.content.split("{SUMMARY}")[1].trim().substring(0, 110)}...`;
};

export const getPostHeaderImage = (markdown: Markdown): string => {
  const headerImageFilePath = path.join(
    markdown.path_to_file,
    markdown.metadata.headerImage
  );
  return imageToBase64.base64Sync(headerImageFilePath) as string;
};

export const getPosts = (): Post[] => {
  return getMarkdown().map((markdown) => {
    const summary = getPostSummary(markdown);
    const header_image = getPostHeaderImage(markdown);
    return {
      ...markdown,
      content: markdown.content.split("{START}")[1].trim(),
      summary,
      header_image,
      pid: seoUrl(markdown.metadata.title),
    };
  });
};

export const mapPidToPostIndex = (posts: Post[]): PidToPostMap => {
  const map: PidToPostMap = {};
  posts.forEach((post, index) => {
    map[post.pid] = index;
  });
  return map;
};

export const allPosts: Post[] = (() => {
  const posts = getPosts();
  posts.sort((postA, postB) =>
    postB.metadata.datePublished.localeCompare(postA.metadata.datePublished)
  );
  return posts;
})();

export const pidToPostIndex: PidToPostMap = mapPidToPostIndex(allPosts);

export default { getMarkdown, getPosts, mapPidToPostIndex, getPostSummary };
