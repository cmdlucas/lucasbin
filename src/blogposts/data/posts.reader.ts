import fs from "fs";
import path from "path";

export interface PostFile {
  directory: string;
  name: string;
  content?: string;
}

/**
 * Get all the posts
 *
 * @return {PostFile[]}
 */
export function readPosts(): PostFile[] {
  /**
   * Using process.pwd() because the directory into which next compiles
   * the code will be different from our current dir resolved using __dirname.
   *
   * See @link https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation
   */
  const postsRootDir = path.join(process.cwd(), "posts");
  const thisDirContent = fs.readdirSync(postsRootDir, { withFileTypes: true });
  const postDirs = thisDirContent.filter((name) => name.isDirectory());
  return postDirs.map((postDir) => {
    const pathToPost = path.join(postsRootDir, postDir.name);
    return { directory: pathToPost, name: "post.md" };
  });
}

/**
 * Get all the posts as individual string content
 *
 * @return {PostFile[]}
 */
export function readPostsContent(): PostFile[] {
  return readPosts().map((postFile: PostFile) => ({
    ...postFile,
    content: fs.readFileSync(
      path.join(postFile.directory, postFile.name),
      "utf8"
    ),
  }));
}
