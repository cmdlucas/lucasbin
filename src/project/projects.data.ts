const Role = {
  SOFTWARE_ENGINEERING: 'Software Engineering',
  CTO: 'CTO',
  CO_FOUNDER: 'Co-Founder',
} as const;

type Roles = (typeof Role)[keyof typeof Role][];

export interface Project {
  title: string;
  description: string;
  roles: Roles;
  linkToProject?: string;
  techStack?: string[];
  desktopScreenshot?: string;
}

const zubbah: Project = {
  title: 'ZUBBAH.COM (C. 2016 - 2017)',
  description: 'Online marketplace for a wide range of products.',
  roles: [Role.SOFTWARE_ENGINEERING],
  linkToProject: 'https://zubbah.com',
};

const jolibizAjent: Project = {
  title: 'JOLIBIZ AGENT (C. 2017 - 2019)',
  description: 'Deals management platform for educational consultants.',
  roles: [Role.SOFTWARE_ENGINEERING],
  linkToProject: 'https://jolibiz.com',
};

const tokeeto: Project = {
  title: 'TOKEETO (C. 2019 - 2020)',
  description: 'Inter-city bus hailing service.',
  roles: [Role.CO_FOUNDER, Role.CTO],
  linkToProject: 'https://tinyurl.com/tokeeto-rider',
};

const jimdo: Project = {
  title: 'JIMDO (C. 2020 - 2022)',
  description: 'Website builder for small businesses.',
  roles: [Role.SOFTWARE_ENGINEERING],
  linkToProject: 'https://jimdo.com',
};

const mero: Project = {
  title: 'MERO (C. 2021 - 2022)',
  description: 'Property cleaning management.',
  roles: [Role.SOFTWARE_ENGINEERING],
  linkToProject: 'https://mero.co',
};

const collectAI: Project = {
  title: 'COLLECT AI (C. 2022 - DATE)',
  description: 'Dunning process management for businesses.',
  roles: [Role.SOFTWARE_ENGINEERING],
  linkToProject: 'https://collect.ai',
};

export const allProjects = [
  collectAI,
  jimdo,
  mero,
  tokeeto,
  jolibizAjent,
  zubbah,
];
