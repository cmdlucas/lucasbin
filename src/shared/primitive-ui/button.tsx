import styled from "styled-components";

export const PrimaryButton = styled.button(props => ({
    border: "none",
    background: props.theme.main.buttonPrimary.background,
    color: props.theme.main.buttonPrimary.color,
    fontSize: "1.25em",
    textTransform: "capitalize",
    minWidth: "160px",
    textAlign: "center",
    padding: "13.5px 20px"
}))

export const SecondaryButton = styled(PrimaryButton)(props => ({
    background: props.theme.main.buttonSecondary.background,
    color: props.theme.main.buttonSecondary.color,
}))