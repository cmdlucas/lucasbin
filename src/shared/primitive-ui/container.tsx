import styled from "styled-components";

export const HMFContainer = styled.div(() => ({
    margin: "auto",
    maxWidth: "1140px",
}))
