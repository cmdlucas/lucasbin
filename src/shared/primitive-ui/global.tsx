import styled, { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
    @font-face {
        font-family: Poppins;
        src: url(/fonts/poppins/poppins-v6-latin-regular.ttf);
    }
    @font-face {
        font-family: Inconsolata;
        src: url(/fonts/inconsolata/Inconsolata-Regular.ttf);
    }
    @font-face {
        font-family: CooperHewitt;
        src: url(/fonts/cooper-hewitt/CooperHewitt-Medium.ttf);
    }
    
    * {
        margin: 0px;
        font-family: Inconsolata
    }

    a, button {
        cursor: pointer;
    }
    
    #nprogress {
        pointer-events: none;
    }

    #nprogress .bar {
        background: #29d;
        position: fixed;
        z-index: 20000;
        top: 0;
        left: 0;
        width: 100%;
        height: 2px;
    }

    /* Fancy blur effect */
    #nprogress .peg {
        display: block;
        position: absolute;
        right: 0px;
        width: 100px;
        height: 100%;
        box-shadow: 0 0 10px #29d, 0 0 5px #29d;
        opacity: 1.0;

        -webkit-transform: rotate(3deg) translate(0px, -4px);
            -ms-transform: rotate(3deg) translate(0px, -4px);
                transform: rotate(3deg) translate(0px, -4px);
    }

    /* Remove these to get rid of the spinner */
    #nprogress .spinner {
        display: block;
        position: fixed;
        z-index: 20000;
        top: 4px;
        right: 15px;
    }

    #nprogress .spinner-icon {
        width: 18px;
        height: 18px;
        box-sizing: border-box;

        border: solid 2px transparent;
        border-top-color: #29d;
        border-left-color: #29d;
        border-radius: 50%;

        -webkit-animation: nprogress-spinner 400ms linear infinite;
        animation: nprogress-spinner 400ms linear infinite;
    }

    .nprogress-custom-parent {
        overflow: hidden;
        position: relative;
    }

    .nprogress-custom-parent #nprogress .spinner,
    .nprogress-custom-parent #nprogress .bar {
        position: absolute;
    }

    @-webkit-keyframes nprogress-spinner {
        0%   { -webkit-transform: rotate(0deg); }
        100% { -webkit-transform: rotate(360deg); }
    }
    @keyframes nprogress-spinner {
        0%   { transform: rotate(0deg); }
        100% { transform: rotate(360deg); }
    }
`

export const AbsolutePosition = styled.div(props => ({
    position: "absolute",
}))

export const RelativePosition = styled.div(props => ({
    position: "relative"
}))

export const FixedPosition = styled.div(props => ({
    position: "fixed"
}))

export const HrLine = styled.hr(props => ({
    border: `1px solid ${props.theme.main.borderColor}`
}))