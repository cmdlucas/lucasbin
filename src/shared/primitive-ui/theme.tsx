export interface ButtonTheme {
  color: string;
  background: string;
}

export interface ThemeProperties {
  textColor: string;
  linkColor: string;
  background: string;
  borderColor: string;
  buttonPrimary: ButtonTheme;
  buttonSecondary: ButtonTheme;
}

export const lightTheme: ThemeProperties = {
  background: "#FFFFFF",
  textColor: "#212424",
  linkColor: "#0D66D0",
  borderColor: "#EDEDED",
  buttonPrimary: {
    color: "#FFFFFF",
    background: "#212424",
  },
  buttonSecondary: {
    color: "#FFFFFF",
    background: "#366DDC",
  },
};

export const darkTheme: ThemeProperties = {
  background: "#212424",
  textColor: "#F3F3F3",
  linkColor: "#CEC652",
  borderColor: "#555555",
  buttonPrimary: {
    color: "#212424",
    background: "#FFFFFF",
  },
  buttonSecondary: {
    color: "#FFFFFF",
    background: "#366DDC",
  },
};

export type ThemeType = "light" | "dark";

export interface Theme {
  type: ThemeType;
  main: ThemeProperties;
  switchTheme: () => void;
}

export const invertTheme = (themeType: ThemeType): ThemeProperties =>
  themeType === "light" ? darkTheme : lightTheme;
  
export const invertThemeType = (themeType: ThemeType): ThemeType =>
  themeType === "light" ? "dark" : "light";

export const defaultTheme: Theme = {
  type: "light",
  main: lightTheme,
  switchTheme: () => {
    return;
  },
};
