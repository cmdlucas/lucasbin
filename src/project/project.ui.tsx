import React, { FC } from 'react';
import { Project } from './projects.data';
import styled from 'styled-components';
import {
  AbsolutePosition,
  RelativePosition,
} from '../shared/primitive-ui/global';
import { HeaderThree, Paragraph, TextLink } from '../shared/primitive-ui/text';
import { FlexRowWrap } from '../shared/primitive-ui/flexbox';

const ImageHolder = styled(RelativePosition)`
  overflow: hidden;
  height: 300px;
  margin-top: 16px;
`;

const Shade = styled(AbsolutePosition)`
  width: 100%;
  height: 100%;
  bottom: 0px;
  left: 0px;
  background: rgba(145, 145, 145, 0.7);
`;

const Description = styled(Paragraph)`
  margin-top: 16px;
`;

const Article = styled.article`
  code {
    background: ${(props) =>
      props.theme.type === 'light' ? '#EDEDED' : '#1E1E1E'};
    padding: 4px 8px;
    color: ${(props) => (props.theme.type === 'light' ? '#666666' : '#ACACAC')};
    border-radius: 4px;
    margin-bottom: 12px;
  }
  p {
    margin: 16px 0px;
  }
  margin-bottom: 24px;
`;

export const OneProject: FC<Project> = (props) => {
  return (
    <Article>
      <HeaderThree> {props.title} </HeaderThree>
      <Description> {props.description} </Description>

      {props.linkToProject && (
        <Description>
          {' '}
          Website:
          <TextLink href={props.linkToProject}>{props.linkToProject}</TextLink>
        </Description>
      )}

      {props.roles && (
        <>
          <Paragraph>Role: </Paragraph>
          <FlexRowWrap>
            {props.roles.map((tech, index) => (
              <code
                style={{ marginLeft: index > 0 ? '12px' : '0px' }}
                key={index}
              >
                {tech}
              </code>
            ))}
          </FlexRowWrap>
        </>
      )}

      {props.desktopScreenshot && (
        <ImageHolder>
          <img height="300px" src={props.desktopScreenshot} />
          <Shade />
        </ImageHolder>
      )}

      {props.techStack && (
        <>
          <Paragraph>Tech stack: </Paragraph>
          <FlexRowWrap>
            {props.techStack.map((tech, index) => (
              <code
                style={{ marginLeft: index > 0 ? '12px' : '0px' }}
                key={index}
              >
                {tech}
              </code>
            ))}
          </FlexRowWrap>
        </>
      )}
    </Article>
  );
};

export default React.memo(OneProject);
