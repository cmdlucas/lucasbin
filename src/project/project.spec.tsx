import React from "react";
import renderer from "react-test-renderer";
import { OneProject } from "./project.ui";
import { allProjects } from "./projects.data";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from '../shared/primitive-ui/theme';

describe("Project Component", () => {
  it("should match contain all the required elements in the tree", () => {
    const tree = renderer.create(
      <ThemeProvider theme={defaultTheme}>
        <OneProject {...allProjects[0]} />
      </ThemeProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
