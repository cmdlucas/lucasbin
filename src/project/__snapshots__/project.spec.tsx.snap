// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Component should match contain all the required elements in the tree 1`] = `
<article
  className="sc-fzoLag kUyYvR"
>
  <h3
    className="sc-AxmLO kYuaTq"
  >
     
    JIMDO (C. 2020 - DATE)
     
  </h3>
  <p
    className="sc-fzpans sc-fzoyAV cWLtkz"
  >
     
    Website builder for small businesses.
     
  </p>
  <p
    className="sc-fzpans sc-fzoyAV cWLtkz"
  >
     Website:
    <a
      className="sc-AxhUy fuzwGB"
      href="https://jimdo.com"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      https://jimdo.com
    </a>
  </p>
  <p
    className="sc-fzpans SrvuB"
  >
    Role: 
  </p>
  <div
    className="sc-fzplWN sc-fznyAO sc-fznKkj kqSeSn"
  >
    <code
      style={
        Object {
          "marginLeft": "0px",
        }
      }
    >
      Software Engineer
    </code>
  </div>
</article>
`;
